Content.Server/Disposal/Unit/EntitySystems/DisposableSystem.cs

```diff
@@ -1,18 +1,185 @@
     internal sealed class DisposableSystem : EntitySystem
     {
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly DisposalUnitSystem _disposalUnitSystem = default!;
+        [Dependency] private readonly DisposalTubeSystem _disposalTubeSystem = default!;
+
+        public void ExitDisposals(EntityUid uid, DisposalHolderComponent? holder = null, TransformComponent? holderTransform = null)
+        {
+            if (!Resolve(uid, ref holder))
+                return;
+            if (!Resolve(uid, ref holderTransform))
+                return;
+
+            // Check for a disposal unit to throw them into and then eject them from it.
+            // *This ejection also makes the target not collide with the unit.*
+            // *This is on purpose.*
+            var grid = _mapManager.GetGrid(holderTransform.GridID);
+            var gridTileContents = grid.GetLocal(holderTransform.Coordinates);
+            DisposalUnitComponent? duc = null;
+            foreach (var contentUid in gridTileContents)
+            {
+                if (EntityManager.TryGetComponent(contentUid, out duc))
+                    break;
+            }
+
+            foreach (var entity in holder.Container.ContainedEntities.ToArray())
+            {
+                if (entity.TryGetComponent(out IPhysBody? physics))
+                {
+                    physics.CanCollide = true;
+                }
+
+                holder.Container.ForceRemove(entity);
+
+                if (entity.Transform.Parent == holderTransform)
+                {
+                    if (duc != null)
+                    {
+                        // Insert into disposal unit
+                        entity.Transform.Coordinates = new EntityCoordinates(duc.OwnerUid, Vector2.Zero);
+                        duc.Container.Insert(entity);
+                    }
+                    else
+                    {
+                        entity.Transform.AttachParentToContainerOrGrid();
+                    }
+                }
+            }
+
+            if (duc != null)
+            {
+                _disposalUnitSystem.TryEjectContents(duc);
+            }
+
+            var atmosphereSystem = EntitySystem.Get<AtmosphereSystem>();
+
+            if (atmosphereSystem.GetTileMixture(holderTransform.Coordinates, true) is {} environment)
+            {
+                atmosphereSystem.Merge(environment, holder.Air);
+                holder.Air.Clear();
+            }
+
+            EntityManager.DeleteEntity(uid);
+        }
+
+        // Note: This function will cause an ExitDisposals on any failure that does not make an ExitDisposals impossible.
+        public bool EnterTube(EntityUid holderUid, EntityUid toUid, DisposalHolderComponent? holder = null, TransformComponent? holderTransform = null, IDisposalTubeComponent? to = null, TransformComponent? toTransform = null)
+        {
+            if (!Resolve(holderUid, ref holder))
```

Comment:
Afaik there are resolve methods with more than one ref argument so you could save a few loc here

Reply:
...are there? ...so there are, need to deal with this

Reply:
Done